the-collab-lab/tcl-54-smart-shopping-list
# Issue 8 | Mark shopping list item as purchased - ac & sj

## Description



### `ListItem.jsx`


`useEffect()` is called any time the user navigates to or refreshes `List` view. For each item in the list, it compares `dateLastPurchased` to `currentDate`.  If the time difference is less than 24 hours, the item will be checked, otherwise the item will remain unchecked.

Additionally, we have a `handleCheck` function to keep track whether an item on the list has been checked. 

If user checks an item:
- We call our `setCheck` function to access the previous state of `check`. At this point, the value is `false` (the default). Then, we call the `updateItem` function from our firebase API. We pass 3 arguments to this function: our token (from local storage), the item id we passed as a prop, and the negation of `prevCheck` -- in other words, `true`.

If user un-checks an item:
- We still have all the same functionality. The difference here is that because the item is un-checked, the value of `prevCheck` would revert to `false`.

### `firebase.js`

We added the `updateItem()` function that is called when a shopping list item is checked. The function takes three parameters: `listId`, `itemId`, and `checked`. First the function gets the value of `totalPurchases` and `dateLastPurchased` from the item. When an item is checked, the `dateLastPurchased` updated to `new Date()`, and `totalPurchases` increases by one. If the user accidentally checks an item and unchecks it, `totalPurchases` decreases by one, and we're still trying to figure out how to change `dateLastPurchased` to its previous timestamp value.

## Related Issue

Closes #8

## Acceptance Criteria


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before
![Screenshot 2023-02-01 205443](https://user-images.githubusercontent.com/102492480/216235517-36073f83-0c6c-4920-ae30-9a242ba2c364.jpg)

### After

Item in list will:
- increment the `totalPurchases` property when checked
- decrement the `totalPurchases` property when un-checked

![ui-issue8](https://user-images.githubusercontent.com/102492480/216782983-6c392754-a616-4f17-8c55-4fcc73209f94.gif)

## Testing Steps / QA Criteria

1. Using your terminal, pull this branch with `git pull origin ac-sj-mark-shopping-list-item`.
2. Use `npm start` to run this branch locally and test the features.
3. Join an existing shopping list (like `my-test-list`) or start a new one and add items if there are none.
4. Navigate to [Firebase](https://console.firebase.google.com/u/0/project/tcl-54-smart-shopping-list/firestore/data/~2Fburl%20swirl%20ahem~2FiPIevLRFXvR3vA4PtuY1) to refer to when completing the next step.
5. In `List` view, check items in the list and look to see that `dateLastPurchased` and `totalPurchases` were updated.

## To-Dos
-  When user un-checks the item (because maybe they checked it by mistake), the `dateLastPurchased` property should revert back to its original state (either null or the last time it was purchased). Currently it is not updating to the expected timestamp/date.
